### PR TITLE
Flyway and Liquibase are now run as init containers in manifests.

### PR DESCRIPTION
**Pull request by iocanel**
The pull request introduces an initialization phase where init tasks like Flyway & Liquibase migration take place. It allows users to run those tasks in isolation (without running the whole application).

For Kubernetes environments where it only makes sense for such tasks to run once, the Kubernetes manifest generator exposes them as `Jobs` and tunes `Deployment` so that they only run once initialization is complete. 

In other words, it enables the scaling of Quarkus applications that use init tasks in Kubernetes.

### CODE REVIEW

**Review by geoand — CHANGES_REQUESTED** [review]
I made a suggestion for properly stopping the application

**Review by geoand — APPROVED** [review]
This is awesome, thanks!

**Comment by geoand** [comment]
We should also include documentation in the liquibase and flyway extensions about this

File: extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java [hunk 1]
```diff
@@ -216,6 +221,18 @@ public ServiceStartBuildItem startActions(FlywayRecorder recorder,
         return new ServiceStartBuildItem("flyway");
     }
 
+    @BuildStep
+    void configureKubernetes(ContainerImageInfoBuildItem imageInfo,
```
Comment (maxandersen) [inline extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java:225]: is there a reason why flyway will have to be dependent on kubernetes for this? 

wouldn't this be possible to be generic and have notion of running init tasks vs running fully?

Comment (iocanel) [reply extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java:225]: All extensions that influence the generation of kubernetes manifests need to depend on the kubernetes spi.
The kubernetes spi is a zero dependency (I think) module that only defines build items realted to the generation of the manfiests.

There is no technical reason for not using something more abstract other than that we don't have something in place.
We could possibly define an abstract Job or Init Task and have it translate to kubernetes, amazon lambda, or maybe in the future tekton etc. I can give it a spin and see how it goes.

**Comment by loicmathieu** [comment]
Liquibase MongoDB is missing, if you didn't have time to include it in this PR can you open an issue to not forget it ?
I didn't had time to look at the PR but this looks like a good idea, I'll read it later when I'll find the time.

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRuntimeConfig.java [hunk 2]
```diff
@@ -21,6 +21,13 @@ public FlywayDataSourceRuntimeConfig getConfigForDataSourceName(String dataSourc
         return namedDataSources.getOrDefault(dataSourceName, FlywayDataSourceRuntimeConfig.defaultConfig());
     }
 
+    /**
+     * Flag to enable / disable Liquibase.
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRuntimeConfig.java:25]:
```suggestion
     * Flag to enable / disable Flyway.
```

Comment (iocanel) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRuntimeConfig.java]: Thanks

**Comment by iocanel** [comment]
Forgot to mention:
- Test
- Docs

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java [hunk 3]
```diff
@@ -119,6 +119,13 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
     @ConfigItem
     public boolean validateAtStart;
 
+    /**
+     * true to quit exit right after the operation.
+     *
+     */
+    @ConfigItem
+    public boolean runAndExit;
```
Comment (geoand) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:127]: Are users meant to set this? If not, we should probably add warning that this is considered an advanced setting.

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRuntimeConfig.java [hunk 4]
```diff
@@ -21,6 +21,13 @@ public FlywayDataSourceRuntimeConfig getConfigForDataSourceName(String dataSourc
         return namedDataSources.getOrDefault(dataSourceName, FlywayDataSourceRuntimeConfig.defaultConfig());
     }
 
+    /**
+     * Flag to enable / disable Flyway.
+     *
+     */
+    @ConfigItem(defaultValue = "true")
+    public boolean enabled;
```
Comment (geoand) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRuntimeConfig.java:29]: When will this be used?

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddJobResourceDecorator2.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
+
+package io.quarkus.kubernetes.deployment;
+
+import java.util.List;
+
+import io.dekorate.kubernetes.decorator.ResourceProvidingDecorator;
+import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
+import io.fabric8.kubernetes.api.model.batch.v1.JobBuilder;
+
+public class AddJobResourceDecorator2 extends ResourceProvidingDecorator<KubernetesListBuilder> {
```
Comment (geoand) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddJobResourceDecorator2.java:10]: `AddJobResourceDecorator2`? Is there a reason for this name?

File: extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java [hunk 6]
```diff
@@ -15,6 +15,13 @@
 @ConfigRoot(name = "liquibase-mongodb", phase = ConfigPhase.RUN_TIME)
 public class LiquibaseMongodbConfig {
 
+    /**
+     * Flag to enable / disable Liquibase.
+     *
+     */
+    @ConfigItem(defaultValue = "true")
+    public boolean enabled;
```
Comment (geoand) [inline extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java:23]: Same question as the Flyway extension

File: extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java [hunk 7]
```diff
@@ -33,6 +40,13 @@ public class LiquibaseMongodbConfig {
     @ConfigItem
     public boolean cleanAtStart;
 
+    /**
+     * true to quit exit right after the operation.
+     *
+     */
+    @ConfigItem
+    public boolean runAndExit;
```
Comment (geoand) [inline extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java:48]: Same question as the Flyway extension

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java [hunk 8]
```diff
@@ -58,6 +58,13 @@ public static final LiquibaseDataSourceRuntimeConfig defaultConfig() {
     @ConfigItem
     public boolean cleanAtStart;
 
+    /**
+     * true to quit exit right after the operation.
+     *
+     */
+    @ConfigItem
+    public boolean runAndExit;
```
Comment (geoand) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java:66]: Same question as the Flyway extension

**Review by geoand — CHANGES_REQUESTED** [review]
This is amazing!

I have a few questions and suggestions

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java [hunk 9]
```diff
@@ -27,9 +28,9 @@ public class FlywayRecorder {
 
     static final List<FlywayContainer> FLYWAY_CONTAINERS = new ArrayList<>(2);
 
-    private final FlywayRuntimeConfig config;
+    private final RuntimeValue<FlywayRuntimeConfig> config;
```
Comment (gsmet) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:31]: Hmmm. This is weird? I don't think you need to wrap this to inject the runtime config. At least I don't remember having to do it.

@geoand do I miss something?

**Review by gsmet — DISMISSED** [review]
I asked something that I would like to be clarified.

Also probably a good idea to at least squash this commit given it's a fixup (and maybe others, I have no idea if they are semantically separate).

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java (hunk 9 above)
Comment (geoand) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:31]: AFAIR, it does need to be wrapped

Comment (iocanel) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:31]: In the Quarkus codebase I've encountered both styles. So, I initially went without wrapping but encountered errors as soon as I added integration tests.

**Comment by iocanel** [comment]
@gsmet: The `fixup` commits are meant to be squashed after the review phase. I can squash them now if this is preferable.

File: core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java [hunk 10]
```diff
@@ -205,7 +203,10 @@ public static void run(Application application, Class<? extends QuarkusApplicati
                 stateLock.unlock();
             }
             application.stop();
-            (exitCodeHandler == null ? defaultExitCodeHandler : exitCodeHandler).accept(1, e);
+            int exceptionExitCode = rootCause instanceof PreventFurtherStepsException
+                    ? ((PreventFurtherStepsException) rootCause).getExitCode()
+                    : 1;
+            (exitCodeHandler == null ? defaultExitCodeHandler : exitCodeHandler).accept(exceptionExitCode, e);
```
[the remark this answers is not in the document]
Comment (iocanel) [reply core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java:209]: The old code used to pass a hard coded exit code value of 1 to the exit code handler (if available).
This change takes into consideration that `PreventFurtherStepsException` may yield a zero exit code value.

Why do we need a zero exit code value ?

A Kubernetes Job to be marked as complete needs to run and return a non zero exit code, or else it is retired.
So, in order to be able to run flyway / liquibase as kubernetes jobs it's needed to be able to exit with a zero exit code.

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java (hunk 3 above)
Comment (iocanel) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:127]: @geoand do we have an example?

File: core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java:209]: Cool, thanks

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java (hunk 3 above)
Comment (geoand) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:127]: Not that I recall. Just add some strongly worded statement since this stuff ends up in generated docs pages so we need to make sure users should not casually use this

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java (hunk 9 above)
Comment (geoand) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:31]: Yeah, I think it depends on timinng, but in general this pattern should be used

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddJobResourceDecorator2.java (hunk 5 above)
Comment (geoand) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddJobResourceDecorator2.java:10]: I guess we should at least have a comment then about the intent

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRuntimeConfig.java (hunk 4 above)
Comment (iocanel) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRuntimeConfig.java:29]: It's used to turn flyway off in the main container, since the flyway init is externalized.

File: extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java (hunk 6 above)
Comment (iocanel) [reply extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java:23]: Acts as a flag to turn the job off, when we externalize it.

File: extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java (hunk 7 above)
Comment (iocanel) [reply extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java:49]: Acts as a flag to turn the job off, when we externalize it.

Comment (iocanel) [reply extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/runtime/LiquibaseMongodbConfig.java:49]: Added comment

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java (hunk 8 above)
Comment (iocanel) [reply extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java:67]: Added comment

**Comment by geoand** [comment]
The test failures are definitely related

File: extensions/kubernetes/spi/src/main/java/io/quarkus/kubernetes/spi/KubernetesJobBuildItem.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
+package io.quarkus.kubernetes.spi;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * A built item for generating init containers.
```
Comment (geoand) [inline extensions/kubernetes/spi/src/main/java/io/quarkus/kubernetes/spi/KubernetesJobBuildItem.java:10]:
```suggestion
 * A Built Item for generating init containers.
```

**Review by geoand — COMMENTED** [review]
I only added a few tiny comments.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/InitTaskBuildItem.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+public final class InitTaskBuildItem extends MultiBuildItem {
+
+    private final String name;
+    private final Optional<String> image;
+    private final List<String> command;
+    private final List<String> arguments;
+    private final Map<String, String> taskEnvVars;
+    private final Map<String, String> appEnvVars;
+    private final boolean sharedEnvironment;
+    private final boolean sharedFilesystem;
```
[the remark this answers is not in the document]
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/InitTaskBuildItem.java:28]: I still have no idea what sharedEnvironment and sharedFilesystem are :)

**Review by geoand — APPROVED** [review]
💪🏼

**Comment by geoand** [comment]
@iocanel can you do a final rebase onto main so we can get this in?

**Comment by iocanel** [comment]
> @iocanel can you do a final rebase onto main so we can get this in?

I intend to add a few more things to the pull request, before we merge it.

**Comment by iocanel** [comment]
@geoand: added a commit that adds global handing of `exit after init`.

File: core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java [hunk 13]
```diff
@@ -107,9 +107,12 @@ public static void run(Application application, Class<? extends QuarkusApplicati
             stateLock.unlock();
         }
         try {
+
+            System.out.println("Calling application start.");
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java:111]: We probably want to get rid of these :)

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/InitalizationCompletedBuildItem.java [hunk 14]
```diff
@@ -0,0 +1,16 @@
+package io.quarkus.deployment.builditem;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class InitalizationCompletedBuildItem extends MultiBuildItem {
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/InitalizationCompletedBuildItem.java:5]: We definitely need some Javadoc on this. Also, I am wondering if the name is too generic... Maybe we need to be specific about what kind of initialization we are talking about.

File: core/runtime/src/main/java/io/quarkus/runtime/init/InitRecorder.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
+package io.quarkus.runtime.init;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.runtime.PreventFurtherStepsException;
+import io.quarkus.runtime.RuntimeValue;
+import io.quarkus.runtime.annotations.Recorder;
+
+@Recorder
+public class InitRecorder {
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkus/runtime/init/InitRecorder.java:10]: Same comment about the name

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/init/InitProcessor.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+import io.quarkus.arc.deployment.SyntheticBeansRuntimeInitBuildItem;
+import io.quarkus.arc.runtime.init.InitRecorder;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.Consume;
+import io.quarkus.deployment.annotations.ExecutionTime;
+import io.quarkus.deployment.annotations.Record;
+import io.quarkus.deployment.builditem.InitalizationCompletedBuildItem;
+
+public class InitProcessor {
```
Comment (geoand) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/init/InitProcessor.java:13]: And here

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/init/InitRecorder.java [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.runtime.PreventFurtherStepsException;
+import io.quarkus.runtime.RuntimeValue;
+import io.quarkus.runtime.annotations.Recorder;
+import io.quarkus.runtime.init.InitRuntimeConfig;
+
+@Recorder
+public class InitRecorder {
```
Comment (geoand) [inline extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/init/InitRecorder.java:11]: Do we need this class?

**Review by geoand — COMMENTED** [review]
I added some comments about the last commit

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/InitalizationCompletedBuildItem.java (hunk 14 above)
Comment (iocanel) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/InitalizationCompletedBuildItem.java:5]: How about `InitializationTaskCompleteBuildItem`?

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/init/InitRecorder.java (hunk 17 above)
Comment (iocanel) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/init/InitRecorder.java:11]: No we don't, I'll delete that

**Comment by iocanel** [comment]
@geoand applied feedbak and renamed classes to `InitializationTaskXXX` to be slightly more descriptive.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/InitalizationTaskCompletedBuildItem.java [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * A symbolic class that represents that an initialization task has been complete.
+ * Similar to {@link ServiceStartBuildItem} but focused on initialization tasks (e.g. db migrations etc).
+ * <p>
+ * The build item is used, so that we can track when all intialization tasks have been completed.
+ */
+public final class InitalizationTaskCompletedBuildItem extends MultiBuildItem {
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/InitalizationTaskCompletedBuildItem.java:11]: I think we need to have some kind of description of when these tasks are meant to run

File: extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java [hunk 19]
```diff
@@ -208,22 +209,23 @@ void createBeans(FlywayRecorder recorder,
...
+        initializationCompleteBuildItem.produce(new InitalizationTaskCompletedBuildItem("flyway"));
         return new ServiceStartBuildItem("flyway");
     }
 
     @BuildStep
     public InitTaskBuildItem configureInitTask() {
         return InitTaskBuildItem.create()
                 .withName("flyway-init")
-                .withTaskEnvVars(Map.of("QUARKUS_FLYWAY_RUN_AND_EXIT", "true", "QUARKUS_FLYWAY_ENABLED", "true"))
+                .withTaskEnvVars(Map.of("QUARKUS_INIT_EXIT", "true", "QUARKUS_FLYWAY_ENABLED", "true"))
```
Comment (geoand) [inline extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java:228]: maybe it should be `QUARKUS_INIT_TASK_EXIT` ?

**Comment by iocanel** [comment]
It seems that the latest change broke something. On it ...

**Comment by geoand** [comment]
Mind rebasing on `main`? I want to make sure the CI failures are not real

**Comment by iocanel** [comment]
> Mind rebasing on `main`? I want to make sure the CI failures are not real

Any idea about the liniting issues? I don't get the error message.  @ebullient

**Comment by iocanel** [comment]
> Mind rebasing on `main`? I want to make sure the CI failures are not real

**Comment by iocanel** [comment]
@gsmet: The requested changes have been applied and CI issues have been addressed.

**Comment by geoand** [comment]
🎉 

I'll have another look tomorrow

File: core/runtime/src/main/java/io/quarkus/runtime/init/InitializationTaskRecorder.java [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+import io.quarkus.runtime.PreventFurtherStepsException;
+import io.quarkus.runtime.annotations.Recorder;
+
+/**
+ * A {@link Recorder} that is used to check if the application should exit once all initialization tasks are completed.
+ */
+@Recorder
+public class InitializationTaskRecorder {
+
+    private static final String QUARKUS_INIT_AND_EXIT = "quarkus.initAndExit";
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkus/runtime/init/InitializationTaskRecorder.java:16]: Minor complaint: This should probably be `quarkus.init-and-exit` if we are to stick with conventions

File: docs/.vale/styles/Quarkus/Spelling.yml [hunk 21]
```diff
@@ -279,7 +285,6 @@ filters:
   - Minishift
   - Mirantis
   - Mockito
-  - MongoDB
```
Comment (geoand) [inline docs/.vale/styles/Quarkus/Spelling.yml:282]: Not sure why this was removed. Did the tooling do that automatically?

**Review by geoand — COMMENTED** [review]
I only added a couple tiny comments, I think this is almost good to go!

File: docs/.vale/styles/Quarkus/Spelling.yml (hunk 21 above)
Comment (iocanel) [reply docs/.vale/styles/Quarkus/Spelling.yml:282]: It wasn't removed. It was replaced by a regex that handles lowercase m and capital d too.

**Comment by iocanel** [comment]
@geoand: feedback applied, can you please have another look?

**Review by geoand — APPROVED** [review]
🎉
